sphinx/tutorial/rose/furthertopics/widget-development.rst

```diff
@@ -0,0 +1,156 @@
+Widget Development
+==================
+
+
+The ``rose edit`` GUI displays configurations using built-in widgets. For more
+complex requirements ``rose edit`` supports custom widgets as plugins.
+
+In this tutorial we will write a custom widget which offers typing suggestions
+when entering usernames.
+
+.. image:: img/rose-edit-custom-widget.png
+   :align: center
+   :width: 450px
+
+.. warning::
+
+   If you find yourself needing to write a custom widget, please contact the
+   Rose team for guidance.
+
+Example
+-------
+
+Create a new rose app by running the following command replacing
+``DIRECTORY`` with the path in which to create the suite::
+
+   rose tutorial widget DIRECTORY
+   cd DIRECTORY
+
+You will now have a rose app which contains the following files:
+
+.. code-block:: none
+
+   DIRECTORY
+   |-- meta
+   |   `-- lib
+   |       `-- python
+   |           `-- widget
+   |               |-- __init__.py
+   |               `-- username.py
+   `-- rose-app.conf
+
+The ``rose-app.conf`` file defines an environment variable called ``USER``:
+
+.. code-block:: rose
+
+   [env]
+   USER=fred
+
+.. _python package: https://docs.python.org/3/tutorial/modules.html#packages
+
+The ``__init__.py`` file is empty - the presence of this file declares the
+``widget`` directory as a `python package`_.
+
+The ``username.py`` file is where we will write our widget.
+
+Initial Code
+^^^^^^^^^^^^
+
+We will start with a slimmed-down copy of the class
+``rose.config_editor.valuewidget.text.RawValueWidget`` which you will find
+in the file ``username.py``. It contains all the API calls you would normally
+ever need.
+
+We are now going to extend the widget to be more useful.
+
+Add a line importing the ``pwd`` package at the top of the file:
+
+.. code-block:: diff
+
+   + import pwd
+
+    import gobject
+    import pygtk
+    pygtk.require('2.0')
+    import gtk
+
+This adds the Python library that we'll use in a minute.
+
+Now we need to create a predictive text model by adding some data to our
+``gtk.Entry`` text widget.
+
+We need to write our method ``_set_completion``, and put it in the main body
+of the class. This will retrieve usernames from the ``pwd.getpwall()``
+function and store them so they can be used by the text widget ``self.entry``.
+
+Add the following method to the ``UsernameValueWidget`` class:
+
+.. code-block:: python
+
+   def _set_completion(self):
+       # Return a predictive text model.
+       completion = gtk.EntryCompletion()
+       model = gtk.ListStore(str)
+       for username in [p.pw_name for p in pwd.getpwall()]:
+           model.append([username])
+       completion.set_model(model)
+       completion.set_text_column(0)
+       completion.set_inline_completion(True)
+       self.entry.set_completion(completion)
+
+We need to make sure this method gets called at the right time, so we add
+the following line to the ``__init__`` method:
```

Comment:
Potential for readers to think this means via the '\_\_init\_\_ py' file, so perhaps affix with 'at the top of the class'.

Reply:
It says "the `__init__`  method" so I think this'll do.